src/useBattery.ts

```diff
@@ -1,56 +1,106 @@
-    on(battery, 'levelchange', onChange);
-    on(battery, 'chargingtimechange', onChange);
-    on(battery, 'dischargingtimechange', onChange);
-  };
-
-  useEffect(() => {
-    (navigator as any).getBattery().then((bat: any) => {
-      if (mounted) {
-        battery = bat;
-        onBattery();
-      }
-    });
-
-    return () => {
-      mounted = false;
-      if (battery) {
-        off(battery, 'chargingchange', onChange);
-        off(battery, 'levelchange', onChange);
-        off(battery, 'chargingtimechange', onChange);
-        off(battery, 'dischargingtimechange', onChange);
-      }
-    };
-  }, []);
+interface BatteryManager extends Readonly<BatteryState>, EventTarget {
+  onchargingchange: () => void;
+  onchargingtimechange: () => void;
+  ondischargingtimechange: () => void;
+  onlevelchange: () => void;
+}
 
-  return state;
+interface NavigatorWithPossibleBattery extends Navigator {
+  getBattery?: () => Promise<BatteryManager>;
+}
+
+const nav: NavigatorWithPossibleBattery = navigator;
+
+type UseBatteryState = BatteryState & {
+  isSupported: boolean;
 };
 
-export default useBattery;
+export default function useBattery() {
+  const [state, setState] = React.useState<UseBatteryState>({
+    isSupported: nav && typeof nav.getBattery !== 'undefined',
+    level: 1,
+    charging: true,
+    dischargingTime: Infinity,
+    chargingTime: 0,
+  });
+  const battery = React.useRef<BatteryManager>();
+  let isMounted = true;
+
+  if (state.isSupported) {
+    const onChange = React.useCallback(() => {
+      if (isMounted && battery.current) {
+        setState({
+          isSupported: true,
+          level: battery.current.level,
+          charging: battery.current.charging,
+          dischargingTime: battery.current.dischargingTime,
+          chargingTime: battery.current.chargingTime,
+        });
+      }
+    }, [setState]);
+
+    const bindBatteryEvents = React.useCallback(
+      (bat: BatteryManager) => {
+        on(bat, BatteryManagerEvents.chargingChange, onChange);
+        on(bat, BatteryManagerEvents.chargingTimeChange, onChange);
+        on(bat, BatteryManagerEvents.dischargingTimeChange, onChange);
+        on(bat, BatteryManagerEvents.levelChange, onChange);
+      },
+      [onChange]
+    );
+
+    const unbindBatteryEvents = React.useCallback(
+      (bat: BatteryManager) => {
+        off(bat, BatteryManagerEvents.chargingChange, onChange);
+        off(bat, BatteryManagerEvents.chargingTimeChange, onChange);
+        off(bat, BatteryManagerEvents.dischargingTimeChange, onChange);
+        off(bat, BatteryManagerEvents.levelChange, onChange);
+      },
+      [onChange]
+    );
+
+    React.useEffect(() => {
+      battery.current && bindBatteryEvents(battery.current);
+
+      return () => {
+        battery.current && unbindBatteryEvents(battery.current);
+      };
+    }, [onChange]);
+
+    React.useEffect(() => {
+      nav.getBattery!().then((bat: BatteryManager) => {
```

Comment:
This could probably use `useAsync`, then you also would not need to track `isMounted` state.

Reply:
I don't want the component to re-render on async state change - only on battery state change.
Furthermore i still have to clear bound event handler on unmount so still have to use useEffect where i can combine all things.
So for this case, in terms of performance optimisation - better not to use useAsync hook (IMO)